### PR TITLE
6D Router Lookahead

### PR DESCRIPTION
**Pull request by amin1377**
In this pull request, the main data structure of the router lookahead is updated to have six dimensions instead of five. The previous dimensions were: [dest. layer][chanx/y][seg_type][width][height].

The new dimensions are: [source layer][chanx/y][seg_type][dest. layer][width][height].

Previously, since the connections were only from OPINs, each layer had a separate router lookahead. In the case of die crossing, the minimum delay of the inter-layer connection was stored and added to the cost to reach the sink on the destination layer from that point.

As development progresses in adding inter-die connections, we needed to update the router lookahead to consider that connections can cross die on switch blocks. To achieve this, an additional dimension is added to the array. Therefore, the first index related to the layer is now the source layer, and the second index related to the layer pertains to the layer on which the sink is located.

### CODE REVIEW

**Comment by amin1377** [comment]
Titan QoR:
Screenshot from 2023-11-23 08-42-25

**Comment by vaughnbetz** [comment]
Slight 1.6% router time increase, but since place time is up .5% and pack time 1.9% this is hopefully machine load noise.  In any case we need it so merging, and Amin is working on optimization of the overall lookahead size and compute time for the future (beyond this PR).

File: vpr/src/route/router_lookahead_map.h [hunk 1]
```diff
@@ -57,8 +55,14 @@ class Cost_Entry {
 
 /* provides delay/congestion estimates to travel specified distances
  * in the x/y direction */
-typedef vtr::NdMatrix<Cost_Entry, 5> t_wire_cost_map; //[0..num_layers][0..1][[0..num_seg_types-1]0..device_ctx.grid.width()-1][0..device_ctx.grid.height()-1]
+// This is a 6D array storing the cost to travel from a node of type CHANX/CHANY to a point that is dx, dy further, and is on the "layer_num" layer.
+// To store this information, the first index is the layer number that the node under consideration is on, the second index represents the type of channel (X/Y)
+// that the node under consideration belongs to, the third is the segment type (specified in the architecture file under the "segmentlist" tag), the fourth is the
+// "layer_num" mentioned above, the fifth is dx, and the last one is dy.
```
Comment (vaughnbetz) [inline vpr/src/route/router_lookahead_map.h:61]: "layer_num" mentioned above --> target "layer_num" mentioned above

File: vpr/src/route/router_lookahead_map_utils.cpp [hunk 2]
```diff
@@ -17,10 +17,17 @@
...
 
-static void dijkstra_flood_to_wires(int itile, RRNodeId inode, util::t_src_opin_delays& src_opin_delays, util::t_src_opin_inter_layer_delays& src_opin_inter_layer_delays, bool is_multi_layer);
+static void dijkstra_flood_to_wires(int itile, RRNodeId inode, util::t_src_opin_delays& src_opin_delays);
 
 static void dijkstra_flood_to_ipins(RRNodeId node, util::t_chan_ipins_delays& chan_ipins_delays);
 
+/**
+ * @brief Return the maximum ptc number of the SOURCE/OPINs of a tile type
+ * @param itile
+ * @return
```
Comment (vaughnbetz) [inline vpr/src/route/router_lookahead_map_utils.cpp:27]: Say what the @return is

File: vpr/src/route/router_lookahead_map_utils.h [hunk 3]
```diff
@@ -294,9 +292,9 @@ typedef std::vector<std::vector<std::vector<t_reachable_wire_inf>>> t_chan_ipins
 /**
  * @brief For each tile, iterate over its OPINs and store which segment types are accessible from each OPIN
  * @param is_flat
- * @return (segments accessible on the same type, segments accessible on other layer)
+ * @return
  */
-std::pair<t_src_opin_delays, t_src_opin_inter_layer_delays> compute_router_src_opin_lookahead(bool is_flat);
+t_src_opin_delays compute_router_src_opin_lookahead(bool is_flat);
```
Comment (vaughnbetz) [inline vpr/src/route/router_lookahead_map_utils.h:297]: Should probably just get this from the router context, but your call.

**Review by vaughnbetz — APPROVED** [review]
Looks fine; a few minor commenting changes etc. to consider. Could be done in a separate PR if you wish or we can merge without waiting for the commenting CI to finish/pass.

File: vpr/src/route/router_lookahead_map_utils.h (hunk 3 above)
Comment (amin1377) [reply vpr/src/route/router_lookahead_map_utils.h:297]: We don't have it in the router context yet.

**Comment by amin1377** [comment]
@vaughnbetz This PR is ready to merge

**Pull request merged**